### PR TITLE
Add PaginatedResultSet::toArray()

### PR DESCRIPTION
**Pull request by markstory**
This was giving me deprecation warnings that I felt were unnecessary toil for developers to fix. I couldn't easily find a rector method that would solve this and I think the ongoing maintenance cost of a one liner is low.

### CODE REVIEW

File: src/Datasource/Paging/PaginatedResultSet.php [hunk 1]
```diff
@@ -63,6 +63,18 @@ public function count(): int
...
+    /**
+     * Get the paginated items as an array.
+     *
+     * This will exhaust the iterator `items`.
+     *
+     * @return array<array-key, T>
+     */
+    public function toArray(): array
+    {
+        return iterator_to_array($this->items());
```
Comment (ADmad) [inline src/Datasource/Paging/PaginatedResultSet.php]: This is a duplication of `jsonSerialize()`, so maybe just call that instead?

Comment (markstory) [reply src/Datasource/Paging/PaginatedResultSet.php]: That works for me.

File: src/Datasource/Paging/PaginatedInterface.php [hunk 2]
```diff
@@ -68,6 +68,15 @@ public function hasPrevPage(): bool;
...
     public function hasNextPage(): bool;
 
+    /**
+     * Get the paginated items as an array.
+     *
+     * This will exhaust the iterator `items`.
+     *
+     * @return array
+     */
+    public function toArray(): array;
```
Comment (LordSimal) [inline src/Datasource/Paging/PaginatedInterface.php]: Isn't that technically a breaking change since this method was added to the interface?

Comment (dereuromark) [reply src/Datasource/Paging/PaginatedInterface.php]: Yeah, should it be an annotation at the top instead?

Comment (markstory) [reply src/Datasource/Paging/PaginatedInterface.php]: :+1: Oops, I'll remove the interface update.

**Comment by LordSimal** [comment]
To clarify: The deprecation can be reproduced via
```
$query = $this->fetchTable('StaffMembers')->find();
$paginator = App::className(
    NumericPaginator::class,
    'Datasource/Paging',
    'Paginator'
);
/** @var NumericPaginator $paginator */
$paginator = new $paginator();
$resultSet = $paginator->paginate($query);
$tmp = $resultSet->toArray();
```
which results in
```
deprecated: 16384 :: Since 5.1.0: Calling `Cake\ORM\ResultSet` methods, such as `toArray()`, on PaginatedResultSet is deprecated. You must call `items()` first (for example, `items()->toArray()`).
```

I haven't ran into this problem because I just pass down the paginated resultset to the view and therefore it uses the `getIterator` method of that class to go through the items.

what would be the reason to manually convert a paginated resultset to an array and not just loop over it normally?

**Comment by markstory** [comment]
> what would be the reason to manually convert a paginated resultset to an array and not just loop over it normally?

Old habits :smile:

**Comment by LordSimal** [comment]
I haven't found it in the Cake 5.0 migration guide but wasn't there a time where it was required to call `->all()` on the paginated resultset? Maybe this was just something we talked about in the past because this is definitely not needed.

Anyway, looks fine to me if this helps "the old habits" to not get in the way 😁

**Pull request merged**